### PR TITLE
fix JENKINS-68826: Exception creating MySQL database dao after upgrading pipeline-maven-plugin to v3.11.0

### PR DESCRIPTION
**Pull request by bguerin**
https://issues.jenkins.io/browse/JENKINS-68826

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/GlobalPipelineMavenConfig.java [hunk 1]
```diff
@@ -543,6 +543,7 @@ public FormValidation doValidateJdbcConnection(
                             }
                         }
                         switch (metaData.getDatabaseMajorVersion()) {
+                            case 14:
```
Comment (bguerin) [inline jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/GlobalPipelineMavenConfig.java:546]: PostgreSQL integration test is done on the "latest" tag of the Docker image, and right now it points to the 14.x branch

File: jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/dao/AbstractPipelineMavenPluginDao.java [hunk 2]
```diff
@@ -664,12 +665,12 @@ protected SortedSet<String> listDownstreamPipelinesBasedOnMavenDependencies(@Non
...
-                "and MAVEN_ARTIFACT.version = ?3 " +
-                "and MAVEN_ARTIFACT.type = ?4 " +
-                "and (MAVEN_ARTIFACT.classifier = ?5 or (MAVEN_ARTIFACT.classifier is null and ?5 is null)) " +
-                "and downstream_job.jenkins_master_id = ?6";
+                "where MAVEN_ARTIFACT.group_id = ? " +
+                "and MAVEN_ARTIFACT.artifact_id = ? " +
+                "and MAVEN_ARTIFACT.version = ? " +
+                "and MAVEN_ARTIFACT.type = ? " +
+                "and (MAVEN_ARTIFACT.classifier = ? or (MAVEN_ARTIFACT.classifier is null and ? is null)) " +
+                "and downstream_job.jenkins_master_id = ?";
```
Comment (bguerin) [inline jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/dao/AbstractPipelineMavenPluginDao.java:673]: this does not pass on PostgreSQL

File: jenkins-plugin/src/main/resources/sql/mysql/13_migration.sql [hunk 3]
```diff
@@ -1,4 +1,4 @@
-DROP INDEX IDX_MAVEN_ARTIFACT;
+DROP INDEX IDX_MAVEN_ARTIFACT on MAVEN_ARTIFACT;
```
Comment (bguerin) [inline jenkins-plugin/src/main/resources/sql/mysql/13_migration.sql:1]: Here the real fix

**Comment by aheritier** [comment]
Damned enforcer rules

**Comment by bguerin** [comment]
> Damned enforcer rules

my bad, I did not build fully before sending

**Comment by aheritier** [comment]
even if I understand very well the need of such rule, some configurations are possibilities are missing and for exemple like here we don't care about versions in the test classpath (at least more than in runtime and esp for slf4j-api) 

Maybe in a different PR we should reviews the deps after disabling the enforcement with something like this in our parent:


```
 <pluginManagement>
            <plugins>
                <plugin>
                    <artifactId>maven-enforcer-plugin</artifactId>
                    <executions>
                        <execution>
                            <id>display-info</id>
                            <configuration>
                                <rules>
                                    <requireUpperBoundDeps>
                                        <level>WARN</level>
                                    </requireUpperBoundDeps>
                                </rules>
                            </configuration>
                        </execution>
                    </executions>
                </plugin>
            </plugins>
        </pluginManagement>
```

others plugins are doing it like : https://github.com/jenkinsci/warnings-ng-plugin/blob/master/plugin/pom.xml#L555-L569

**Review by aheritier — APPROVED** [review]
It's a really good thing to switch to test containers
We could potentially support more databases and several versions of DBs (if it makes sense - maybe using parametrised test after switching to JUnit 5).

**Comment by bguerin** [comment]
> It's a really good thing to switch to test containers We could potentially support more databases and several versions of DBs (if it makes sense - maybe using parametrised test after switching to JUnit 5).

I am also wondering about using [liquibase](https://www.liquibase.org/) to manage database migrations

**Comment by aheritier** [comment]
Yes using liquibase is a good choice. We use it on some apps with @PierreBtz (i know @recampbell will like too 😁)

**Pull request merged**